TauAnalysis/MCEmbeddingTools/plugins/EmbeddingLHEProducer.cc

```diff
@@ -424,6 +430,50 @@ void EmbeddingLHEProducer::rotate180(TLorentzVector &positiveLepton, TLorentzVec
   return;
 }
 
+void EmbeddingLHEProducer::InitialRecoCorrection(TLorentzVector &positiveLepton, TLorentzVector &negativeLepton) {
+  if (!InitialRecoCorrection_)
+    return;
+  edm::LogInfo("TauEmbedding") << "Applying initial reconstruction correction";
+  TLorentzVector Z = positiveLepton + negativeLepton;
+
+  edm::LogInfo("TauEmbedding") << "MuMinus before. Pt: " << negativeLepton.Pt() << " Mass: " << negativeLepton.M();
+  // std::cout << " MuMinus before. Pt: " << negativeLepton.Pt() << " Mass: " << negativeLepton.M() << " Energy: " << negativeLepton.E() << std::endl;
+  float diLeptonMass = (positiveLepton + negativeLepton).M();
+  if (diLeptonMass > 60. && diLeptonMass < 122.) {
+    // std::cout << "DiLeptonMass " << diLeptonMass << std::endl;
+    float zmass = 91.1876;
+    float correction_deviation =
+        5.;  // to ensure only a correction that drops corresponding to a Gaussian with mean zmass and std. dev. 5 GeV
+    double EmbeddingCorrection =
+        1.138;  // value derived by function fitting to fold embedded mass spectrum back to original selection when using mu -> mu embedding
+    EmbeddingCorrection =
+        EmbeddingCorrection /
+        (EmbeddingCorrection -
+         (EmbeddingCorrection - 1.) * exp(-pow((diLeptonMass - zmass), 2.) / (2. * pow(correction_deviation, 2.))));
+    EmbeddingCorrection = ((diLeptonMass + (EmbeddingCorrection - 1.) * zmass) / (diLeptonMass * EmbeddingCorrection));
+    double correctedpositiveLeptonEnergy = std::sqrt(
+        muonMass_ * muonMass_ + EmbeddingCorrection * positiveLepton.Px() * EmbeddingCorrection * positiveLepton.Px() +
+        EmbeddingCorrection * positiveLepton.Py() * EmbeddingCorrection * positiveLepton.Py() +
+        EmbeddingCorrection * positiveLepton.Pz() * EmbeddingCorrection * positiveLepton.Pz());
+    double correctednegativeLeptonEnergy = std::sqrt(
+        muonMass_ * muonMass_ + EmbeddingCorrection * negativeLepton.Px() * EmbeddingCorrection * negativeLepton.Px() +
+        EmbeddingCorrection * negativeLepton.Py() * EmbeddingCorrection * negativeLepton.Py() +
+        EmbeddingCorrection * negativeLepton.Pz() * EmbeddingCorrection * negativeLepton.Pz());
+    positiveLepton.SetPxPyPzE(EmbeddingCorrection * positiveLepton.Px(),
```

Comment:
If one factors out `EmbeddingCorrection` from the corrected energy then one could just scale the entire 4-vector by this after setting it.

Reply:
Done in a8eae9f. See above review.